bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/ToolBar.java

```diff
@@ -1752,37 +1754,33 @@ private static void handleDPIChange(Widget widget, int newZoom, float scalingFac
 	var seperatorWidth = new int[toolItems.length];
 	int itemCount = toolItems.length;
 
+	record ToolItemData(ToolItem toolItem, TBBUTTON button) {
+	}
+
 	// Remove and re-add all button the let Windows resize the tool bar
-	ToolItem[] items = new ToolItem[itemCount];
-	TBBUTTON[] buttondata = new TBBUTTON[itemCount];
+	Map<Integer, ToolItemData> buttondata = new TreeMap<>();
 	for (int i = itemCount - 1; i >= 0; i--) {
-		ToolItem item = toolItems[i];
+		TBBUTTON lpButton = new TBBUTTON ();
+		OS.SendMessage (toolBar.handle, OS.TB_GETBUTTON, i, lpButton);
+		ToolItem item = toolItems[lpButton.idCommand];
 		if ((item.style & SWT.SEPARATOR) != 0 && item.getControl() != null) {
 			// Take note of widths of separators with control, so they can be resized
 			// at the end
 			seperatorWidth[i] = item.getWidth();
 		}
 		DPIZoomChangeRegistry.applyChange(item, newZoom, scalingFactor);
-
-		TBBUTTON lpButton = new TBBUTTON ();
-		OS.SendMessage (toolBar.handle, OS.TB_GETBUTTON, i, lpButton);
-		buttondata[lpButton.idCommand] = lpButton;
-		items[lpButton.idCommand] = item;
+		buttondata.put(i, new ToolItemData(item, lpButton));
 		OS.SendMessage(toolBar.handle, OS.TB_DELETEBUTTON, i, 0);
 	}
-
 	// Refresh the image lists so the image list for the correct zoom is used
 	toolBar.setImageList(toolBar.getImageList());
 	toolBar.setDisabledImageList(toolBar.getDisabledImageList());
 	toolBar.setHotImageList(toolBar.getHotImageList());
-
 	OS.SendMessage(toolBar.handle, OS.TB_BUTTONSTRUCTSIZE, TBBUTTON.sizeof, 0);
-	for (int i = 0; i < buttondata.length; i++) {
-		TBBUTTON button = buttondata[i];
-		if (button != null) {
-			OS.SendMessage(toolBar.handle, OS.TB_ADDBUTTONS, 1, button);
-			ToolItem item = items[i];
-
+	for (ToolItemData itemData : buttondata.values()) {
```

Comment:
This will not result in the correct order, will it? There is no guarantee that the `values()` are iterated in an order conforming to a standard comparison for the according keys. If I understand correctly, the keys of the map are not even used except for ordering. So couldn't you simply use a LIFO structure (like a stack) in which you put the ToolItemData to achieve exactly what is intended here?

Reply:
Actually I would read the Javadoc like is is doing that "The collection's iterator returns the values in ascending order of the corresponding keys". Nevertheless, as it is really pure LIFO, Stack sounds like the better approach. I updated the PR accordingly.

Reply:
That's interesting. I didn't expect that behavior of a map, but I understand it the same way as you do. So the previous implementation was probably correct, still a stack seems to reflect the intention even better. Thank you for sharing the insights 🙂